File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/algebra/solver/ConstantMeetVariableTerm.java
package edu.cornell.cs.apl.viaduct.algebra.solver;

import com.google.auto.value.AutoValue;
import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;
import java.util.Map;

/** Meet of a constant element and a variable. */
@AutoValue
abstract class ConstantMeetVariableTerm<A extends HeytingAlgebra<A>> implements LeftHandTerm<A> {
  static <A extends HeytingAlgebra<A>> ConstantMeetVariableTerm<A> create(
      A lhs, VariableTerm<A> rhs) {
    return new AutoValue_ConstantMeetVariableTerm<>(lhs, rhs);
  }

  protected abstract A getLhs();

  protected abstract VariableTerm<A> getRhs();

  @Override
  public A getValue(Map<VariableTerm<A>, A> solution) {
    return getLhs().meet(getRhs().getValue(solution));
  }

  @Override
  public final AtomicTerm<A> getNode() {
    return getRhs();
  }

  @Override
  public final DataFlowEdge<A> getInEdge() {
    return new ImplyEdge<>(getLhs());
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/algebra/solver/ConstantMeetVariableTerm.kt
package edu.cornell.cs.apl.viaduct.algebra.solver

import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge

/** Meet of a constant element and a variable. */
internal data class ConstantMeetVariableTerm<A : HeytingAlgebra<A>>(val lhs: A, val rhs: VariableTerm<A>) :
    LeftHandTerm<A> {
    override fun getValue(solution: ConstraintSolution<A>): A {
        return lhs.meet(rhs.getValue(solution))
    }

    override val node: AtomicTerm<A>
        get() = rhs

    override val inEdge: DataFlowEdge<A>
        get() = ImplyEdge(lhs)
}
